app.ts

```diff
@@ -0,0 +1,147 @@
+    const outputArray = JSON.parse(textString);
+    return outputArray;
+}
+
+class tableGenerator {
+    element: HTMLElement;
+    myTable: HTMLElement;
+    firstRecord: number;
+    endRecord: number;
+
+    constructor(element: HTMLElement, firstRecord: number) {
+        this.element = element;
+        this.myTable = document.createElement('table');
+        this.element.appendChild(this.myTable);
+        this.firstRecord = firstRecord;
+    }
+    generateTable() {
+        // get table height based on  remaining window size after buttons
+        const contentHeight = (window.innerHeight - CONTROLS_HEIGHT);
+        this.myTable.setAttribute("height", "contentHeight");
+
+        // calculate row number based on table header + row heights + borders in CSS and ensure its never less than 1
+        let numRows: number = Math.floor((contentHeight - CONTROLS_HEIGHT - 26) / 46);
+        numRows = numRows <= 0 ? 1 : numRows;
+
+        // set number of records to get from server and ensure that the last record doesnt exceed the limit.
+        this.endRecord = this.firstRecord - 1 + numRows;
+        this.endRecord = (this.endRecord >= RECORD_COUNT ? RECORD_COUNT - 1 : this.endRecord);
+
+        // get data from server
+        const columnsString: string = getData("http://localhost:2050/columns");
+        const recordsText: string = getData("http://localhost:2050/records?from=" + this.firstRecord + "&to=" + this.endRecord);
+
+        // parse text from api call into arrays
+        const columnsArray = parseText(columnsString);
+        const recordsTable = parseText(recordsText);
+
+        // write html for table header
+        for (const i of columnsArray) {
+            const th = document.createElement('th');
+            this.myTable.appendChild(th);
+
+            th.appendChild(document.createTextNode(" " + i));
+        }
+
+        // write html for table
+        for (const iEntry of recordsTable) {
+
+            const tr = document.createElement('tr');
+            this.myTable.appendChild(tr);
+
+            for (const jEntry of iEntry) {
+
+                const td = document.createElement('td');
+                td.appendChild(document.createTextNode(" " + jEntry));
+                tr.appendChild(td);
+            }
+        }
+    }
+}
+
+function createPage(firstRecord: number) {
+
+    // ensure that the user input is a number and within range of available records
+    if (firstRecord >= RECORD_COUNT || firstRecord < 0 || isNaN(firstRecord)) {
+        // if (firstRecord >= RECORD_COUNT || firstRecord < 0) {
+        alert('please select a NUMBER between 0 and ' + (RECORD_COUNT - 1));
+        return;
+    }
+
+    const contentsDiv = this.content;
+    contentsDiv.innerHTML = ''; // empty table div to fill in a new table
+    const tableElement = new tableGenerator(contentsDiv, firstRecord); // create table using the tableGenerator class
+    tableElement.generateTable();
+    LAST_ENTRY = tableElement.endRecord;
```

Comment:
Feels like we're duplicating state here, reckon if you move the `createPage` and `nextPage` methods into the `tableGenerator` class, you could get away with only using the `firstRecord` and `endRecord` members.